lib/ABI/FunctionType.cpp

```diff
@@ -388,210 +417,447 @@ class ConversionHelper {
+                           const model::Register::Values PointerToCopyLocation,
+                           const model::Binary &Binary,
+                           TypeVector &TypeDependencies) {
+    if (Used.size() == 0) {
+      // Short-circuit the functions that don't return anything (`void`).
+      return ensurePrimitiveTypeIsAvailable(model::PrimitiveTypeKind::Void,
+                                            0,
+                                            Binary,
+                                            TypeDependencies);
+    }
+
+    if (Allowed.empty()) {
+      // A return value is expected on an ABI that doesn't dedicate any
+      // registers to be used for those.
+      return std::nullopt;
     }
 
-    if (UsedRegisters.size() == 1) {
-      if (UsedRegisters.begin()->Location == PointerToCopyLocation) {
-        if constexpr (DryRun)
-          return model::QualifiedType{};
-        else
-          return getTypeOrDefault(UsedRegisters.begin()->Type,
-                                  PointerToCopyLocation,
-                                  TheBinary);
+    // Special case where the only used register is the pointer-to-copy location
+    if (Used.size() == 1 && Used.begin()->Location == PointerToCopyLocation) {
+      if (Used.begin()->Type.isPointer()) {
+        // If the type is already a pointer, preserve it.
+        return Used.begin()->Type;
       } else {
-        if constexpr (RegisterCount == 0)
-          return std::nullopt;
-        if (AllowedRegisters.front() == UsedRegisters.begin()->Location) {
-          if constexpr (DryRun)
-            return model::QualifiedType{};
-          else
-            return getTypeOrDefault(UsedRegisters.begin()->Type,
-                                    UsedRegisters.begin()->Location,
-                                    TheBinary);
-        } else {
-          return std::nullopt;
-        }
+        // Otherwise return a `void *` pointer as the return value type.
+        static constexpr auto VoidKind = model::PrimitiveTypeKind::Void;
+        auto Result = ensurePrimitiveTypeIsAvailable(VoidKind,
+                                                     0,
+                                                     Binary,
+                                                     TypeDependencies);
+        using model::Qualifier;
+        Result.Qualifiers.emplace_back(Qualifier::createPointer(Architecture));
+        return Result;
+      }
+    }
+
+    // TODO: there are some cases where preserving the total structure of
+    // the return value can be benefitial, this needs more work.
+    // For now, just determine the total size of the return value and if it
+    // fits into a primitive value - emit it, otherwise show those values as
+    // a struct (/note: positional ABIs need a whole lot of extra love in this
+    // regard).
+    llvm::SmallVector<std::optional<model::QualifiedType>, 4> ResultingTypes;
+    bool MustUseTheNextOne = false;
```

Comment:
```suggestion
    bool MustUseAllRemaining = false;
```

Reply:
All this stuff is go-o-o-o-one 🎉 

Luckily the `RegisterStateDeductions` are now available (unlike when this was first written) and I can just rely on it for the dirty work.